rfcs/0008-microtvm-project-api.md

```diff
@@ -0,0 +1,509 @@
+        - Makes for two RPC implementations checked-in to TVM.
+
+2. gRPC
+    Pros:
+        - Widely supported, compact encoding
+        - Clearly-documented API and good support forums
+        - Supports streaming, the most natural way to forward TVM RPC traffic.
+    Cons:
+        - Requires the use of another Python package
+        - Requires the use of an IDL compiler
+        - Intended use case (datacenter-scale RPC) is overkill.
+        - Makes for two RPC implementations checked-in to TVM.
+
+3. TVM RPC
+    Pros:
+        - Already exists in TVM
+        - Some prior art for session forwarding
+        - Binary encoding
+    Cons:
+        - Binary encoding
+        - Impossible to use today without compiling TVM
+        - Implementation is designed around TVM's remote inference use cases, and will
+          likely change as new demands arise there.
+
+TVM RPC was decided against given the requirement that TVM must be compiled. gRPC was considered,
+but ultimately rejected because JSON-RPC can be implemented in a single Python file without adding
+the complexities of an IDL compiler.
+
+## Transport functions
+
+When generating projects that perform host-driven inference or autotuning, TVM needs some way to
+communicate with the project's microTVM RPC server. Prior to this RFC, TVM included driver code for
+various transports (e.g. stdio, PySerial, etc). The Project API places this functionality in the
+API server, so that TVM doesn't need to include any transport-specific dependencies (e.g. PySerial)
+in its required Python dependencies.
+
+There are a couple of subtle details that were changed when re-implementing this interface in
+Project API to reduce the complexity of Project API servers.
+
+### Encapsulating binary data
+
+First, JSON-RPC is an ASCII protocol and as such, binary data can't be transmitted without adding
+escape characters. To avoid unreadable and large payloads over the Project API RPC, some encoding
+scheme needed to be chosen in order to encapsulate the binary data in the protocol. The desired
+properties of the encoding scheme are:
+
+ - Representable in JSON without the need for escapes
+ - Compact given the above constraint
+ - Easy to encode and decode in languages likely to be used for the API server
+
+Another common place this occurs is in transmitting binary data to and from websites, where an
+ASCII alphabet is chosen to represent binary data, and the binary data is translated into the
+alphabet. Since there aren't enough ASCII characters to represent all 2^8 == 256 binary values in
+one byte, a smaller alphabet is chosen, typically with 64 or 85 characters. This is referred to as
+`base64` or `base85`, and the binary data is encoded by modular arithmetic into the smaller
+alphabet. Python provides standard support for these via the `base64` module, so the most compact
+encoding (`base85`) was chosen from those standards to encode binary data in the Project API.
```

Comment:
TVM uses base64 for encoding data in the relay text format. Maybe we should be consistent.

Reply:
that is also technically an easier modulo to compute, i think. would be open to this.

Reply:
Lets just do base64 to be consistent

Reply:
okay fine :)